src/epik_eval/dataloader.py

```diff
@@ -0,0 +1,205 @@
+                _input.append(row['question'])
+                target.append(row['answer'])
+
+            _input = tokenizer(_input, padding=True, return_tensors='pt', return_length=True)
+            self.attention_mask = _input.attention_mask
+            self.input_length = _input.length
+            self.input = _input.input_ids
+            target = tokenizer(target, padding=True, return_tensors='pt', return_length=True)
+            self.target_length = target.length
+            self.target = target.input_ids
+            self.target[self.target == tokenizer.pad_token_id] = -100  # -100 is torch loss' default ignore index
+
+    def __iter__(self):
+        # Accelerate dataloader expects dataset to yield individual samples.
+        # Alternative would be to modify Accelerate's dataloader.
+
+        random_order = list(range(self.input.shape[0]))
+        random.shuffle(random_order)
+
+        if self.model_type == 'causal':
+            for i in random_order:
+                yield self.input[i], self.attention_mask[i], self.input_length[i]
+        elif self.model_type == 't5':
+            for i in random_order:
+                yield self.input[i], self.attention_mask[i], self.input_length[i], self.target[i], self.target_length[i]
+
+    def __len__(self):
+        return self.input.shape[0]
+
+
+class EvaluationDataset(torch.utils.data.IterableDataset):
+    """
+    Args:
+        tokenizer (transformers.PreTrainedTokenizer): Huggingface tokenizer to tokenize samples.
+        batch_size (int): Evaluation batch size
+        model_type (str): 'causal' or 't5'. Default: 'causal'
+        segmented_docs (bool): Whether to use segmented stories. If False, will use unsegmented stories. Default: False
+        unsegmented_document_directory (str): Path to the directory containing the unsegmented stories. Default: 'data/dataset/unsegmented_documents/'
+        segmented_document_directory (str): Path to the directory containing the segmented stories. Default: 'data/dataset/segmented_documents/'
+        eval_questions_csv (str): Path to the evaluation (val & test) questions. Default: 'data/dataset/eval_questions.csv'
+    """
+
+    def __init__(
+            self,
+            tokenizer,
+            batch_size,
+            model_type='causal',
+            segmented_docs=False,
+            unsegmented_document_directory='data/dataset/unsegmented_documents/',
+            segmented_document_directory='data/dataset/segmented_documents/',
+            eval_questions_csv='data/dataset/eval_questions.csv'):
+        super().__init__()
+        assert model_type in ['causal', 't5'], f"Expected model_type to be either 'causal' or 't5', but got: {model_type}"
+
+        qa_df = pd.read_csv(eval_questions_csv, dtype=str, delimiter='|')
+        questions = qa_df['question'].to_list()
+        answers = qa_df['answer'].to_list()
+
+        # Model is also evaluated for its ability to memorize training samples.
+        train_doc_ids = []
```

Comment:
It's suspicious that these are called train when it's an eval dataloader.

Reply:
During evaluation, I also measure how well the model memorized training samples.